nautobot/core/models/querysets.py

```diff
@@ -11,7 +11,12 @@ def count_related(model, field):
     Return a Subquery suitable for annotating a child object count.
     """
     subquery = Subquery(
-        model.objects.filter(**{field: OuterRef("pk")}).order_by().values(field).annotate(c=Count("*")).values("c")
+        model.objects.filter(**{field: OuterRef("pk")})
```

Comment:
This works but it doesn't highlight the root cause of the issue which (I think) is the tree fields being included in the query. Take a look at these SQL queries generated by the old code:

```
>>> Manufacturer.objects.annotate(inventory_item_count=count_related(InventoryItem, "manufacturer")).query.__str__()
'SELECT "dcim_manufacturer"."id", "dcim_manufacturer"."created", "dcim_manufacturer"."last_updated", "dcim_manufacturer"."_custom_field_data", "dcim_manufacturer"."name", "dcim_manufacturer"."description", COALESCE((\n    WITH RECURSIVE __tree (\n        "tree_depth",\n        "tree_path",\n        "tree_ordering",\n        "tree_pk"\n    ) AS (\n        SELECT\n            0 AS tree_depth,\n            array[T.id] AS tree_path,\n            array["_name"]::text[] AS tree_ordering,\n            T."id"\n        FROM dcim_inventoryitem T\n        WHERE T."parent_id" IS NULL\n\n        UNION ALL\n\n        SELECT\n            __tree.tree_depth + 1 AS tree_depth,\n            __tree.tree_path || T.id,\n            __tree.tree_ordering || "_name"::text,\n            T."id"\n        FROM dcim_inventoryitem T\n        JOIN __tree ON T."parent_id" = __tree.tree_pk\n    )\n    SELECT COUNT(*) AS "c" FROM "dcim_inventoryitem" U0 , "__tree" WHERE (U0."manufacturer_id" = "dcim_manufacturer"."id" AND (__tree.tree_pk = U0.id)) GROUP BY U0."manufacturer_id", ("__tree".tree_ordering) ORDER BY ("__tree".tree_ordering) ASC), 0) AS "inventory_item_count" FROM "dcim_manufacturer" ORDER BY "dcim_manufacturer"."name" ASC'
>>> Manufacturer.objects.annotate(inventory_item_count=count_related(DeviceType, "manufacturer")).query.__str__()
'SELECT "dcim_manufacturer"."id", "dcim_manufacturer"."created", "dcim_manufacturer"."last_updated", "dcim_manufacturer"."_custom_field_data", "dcim_manufacturer"."name", "dcim_manufacturer"."description", COALESCE((SELECT COUNT(*) AS "c" FROM "dcim_devicetype" U0 WHERE U0."manufacturer_id" = "dcim_manufacturer"."id" GROUP BY U0."manufacturer_id"), 0) AS "inventory_item_count" FROM "dcim_manufacturer" ORDER BY "dcim_manufacturer"."name" ASC'
```

I tried adding this code:

```py
    manager = model.objects
    if hasattr(model.objects, "without_tree_fields"):
        manager = manager.without_tree_fields()
    subquery = Subquery(
        manager.filter(**{field: OuterRef("pk")}).order_by().values(field).annotate(c=Count("*")).values("c")
    )

    return Coalesce(subquery, 0)
```

Now those queries look like this:

```
>>> Manufacturer.objects.annotate(inventory_item_count=count_related(InventoryItem, "manufacturer")).query.__str__()
'SELECT "dcim_manufacturer"."id", "dcim_manufacturer"."created", "dcim_manufacturer"."last_updated", "dcim_manufacturer"."_custom_field_data", "dcim_manufacturer"."name", "dcim_manufacturer"."description", COALESCE((SELECT COUNT(*) AS "c" FROM "dcim_inventoryitem" U0 WHERE U0."manufacturer_id" = "dcim_manufacturer"."id" GROUP BY U0."manufacturer_id"), 0) AS "inventory_item_count" FROM "dcim_manufacturer" ORDER BY "dcim_manufacturer"."name" ASC'
>>> Manufacturer.objects.annotate(inventory_item_count=count_related(DeviceType, "manufacturer")).query.__str__()
'SELECT "dcim_manufacturer"."id", "dcim_manufacturer"."created", "dcim_manufacturer"."last_updated", "dcim_manufacturer"."_custom_field_data", "dcim_manufacturer"."name", "dcim_manufacturer"."description", COALESCE((SELECT COUNT(*) AS "c" FROM "dcim_devicetype" U0 WHERE U0."manufacturer_id" = "dcim_manufacturer"."id" GROUP BY U0."manufacturer_id"), 0) AS "inventory_item_count" FROM "dcim_manufacturer" ORDER BY "dcim_manufacturer"."name" ASC'
```

What do you think about using my patch that removes the tree fields? Those shouldn't be needed here just to generate a count?

Reply:
Yeah I will add this patch to the code. Thanks for looking into this.